crypto/evp_extra/p_kem.c

```diff
@@ -55,6 +93,77 @@ static int pkey_kem_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY *pkey) {
   return 1;
 }
 
+static int pkey_kem_encapsulate_deterministic(EVP_PKEY_CTX *ctx,
+                                              uint8_t *ciphertext,
+                                              size_t  *ciphertext_len,
+                                              uint8_t *shared_secret,
+                                              size_t  *shared_secret_len,
+                                              const uint8_t *seed,
+                                              size_t *seed_len) {
+  KEM_PKEY_CTX *dctx = ctx->data;
+  const KEM *kem = dctx->kem;
+  if (kem == NULL) {
+    if (ctx->pkey == NULL) {
+      OPENSSL_PUT_ERROR(EVP, EVP_R_NO_PARAMETERS_SET);
+      return 0;
+    }
+    kem = KEM_KEY_get0_kem(ctx->pkey->pkey.kem_key);
+  }
+
+  // Caller is getting parameter values.
+  if (ciphertext == NULL && shared_secret == NULL && seed == NULL) {
+    *ciphertext_len = kem->ciphertext_len;
+    *shared_secret_len = kem->shared_secret_len;
+    *seed_len = kem->encaps_seed_len;
+    return 1;
+  }
+
+  // If not getting parameter values, then both
+  // output buffers need to be valid (non-NULL)
+  if (ciphertext == NULL || shared_secret == NULL || seed == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_MISSING_PARAMETERS);
+    return 0;
+  }
+
+  // The output buffers need to be large enough.
+  if (*ciphertext_len < kem->ciphertext_len ||
+      *shared_secret_len < kem->shared_secret_len) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_BUFFER_TOO_SMALL);
+    return 0;
+  }
+
+  if (*seed_len < kem->encaps_seed_len) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_BUFFER_TOO_SMALL);
+    return 0;
+  }
+
+  // Check that the context is properly configured.
+  if (ctx->pkey == NULL ||
+      ctx->pkey->pkey.kem_key == NULL ||
+      ctx->pkey->type != EVP_PKEY_KEM) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_OPERATON_NOT_INITIALIZED);
+    return 0;
+  }
+
+  // Check that the key has a public key set.
+  KEM_KEY *key = ctx->pkey->pkey.kem_key;
+  if (key->public_key == NULL) {
+    OPENSSL_PUT_ERROR(EVP, EVP_R_NO_KEY_SET);
+    return 0;
+  }
+
+  if (!kem->method->encaps_deterministic(ciphertext, shared_secret, key->public_key, seed)) {
+    return 0;
+  }
+
+  // The size of the ciphertext and the shared secret
+  // that has been writen to the output buffers.
+  *ciphertext_len = kem->ciphertext_len;
+  *shared_secret_len = kem->shared_secret_len;
+
```

Comment:
Do you think it makes sense to update the seed length? If someone passes in more than is needed we will only "use" the first part.

Reply:
To be consistent with the API documentation (e.g., `\*  IN  *\` vs `\*  OUT  *\`) we mark this argument as input (with the exception that a null value means the caller is getting the size).